src/api/tag/controller.go

```diff
@@ -0,0 +1,242 @@
+	ierror "github.com/goharbor/harbor/src/internal/error"
+	"github.com/goharbor/harbor/src/pkg/art"
+	"github.com/goharbor/harbor/src/pkg/artifact"
+	"github.com/goharbor/harbor/src/pkg/immutabletag/match"
+	"github.com/goharbor/harbor/src/pkg/immutabletag/match/rule"
+	"github.com/goharbor/harbor/src/pkg/q"
+	"github.com/goharbor/harbor/src/pkg/signature"
+	"github.com/goharbor/harbor/src/pkg/tag"
+	model_tag "github.com/goharbor/harbor/src/pkg/tag/model/tag"
+	"time"
+)
+
+var (
+	// Ctl is a global tag controller instance
+	Ctl = NewController()
+)
+
+// Controller manages the tags
+type Controller interface {
+	// Ensure
+	Ensure(ctx context.Context, repositoryID, artifactID int64, name string) error
+	// Count returns the total count of tags according to the query.
+	Count(ctx context.Context, query *q.Query) (total int64, err error)
+	// List tags according to the query
+	List(ctx context.Context, query *q.Query, option *Option) (tags []*Tag, err error)
+	// Get the tag specified by ID
+	Get(ctx context.Context, id int64, option *Option) (tag *Tag, err error)
+	// Create the tag and returns the ID
+	Create(ctx context.Context, tag *Tag) (id int64, err error)
+	// Update the tag. Only the properties specified by "props" will be updated if it is set
+	Update(ctx context.Context, tag *Tag, props ...string) (err error)
+	// Delete the tag specified by ID with limitation check
+	Delete(ctx context.Context, id int64) (err error)
+	// DeleteTags deletes all tags
+	DeleteTags(ctx context.Context, tags []*Tag) (err error)
+}
+
+// NewController creates an instance of the default repository controller
+func NewController() Controller {
+	return &controller{
+		tagMgr:       tag.Mgr,
+		artMgr:       artifact.Mgr,
+		immutableMtr: rule.NewRuleMatcher(),
+	}
+}
+
+type controller struct {
+	tagMgr       tag.Manager
+	artMgr       artifact.Manager
+	immutableMtr match.ImmutableTagMatcher
+}
+
+// Ensure ...
+func (c *controller) Ensure(ctx context.Context, repositoryID, artifactID int64, name string) error {
+	query := &q.Query{
+		Keywords: map[string]interface{}{
+			"repository_id": repositoryID,
+			"name":          name,
+		},
+	}
+	tags, err := c.List(ctx, query, &Option{
+		WithImmutableStatus: true,
+		WithSignature:       true,
+	})
+	if err != nil {
+		return err
+	}
+	// the tag already exists under the repository
+	if len(tags) > 0 {
+		tag := tags[0]
+		// existing tag must check the immutable status and signature
+		if tag.Immutable {
+			return ierror.New(nil).WithCode(ierror.PreconditionCode).
+				WithMessage("the tag %s configured as immutable, cannot be updated", tag.Name)
+		}
+		if tag.Signed {
+			return ierror.New(nil).WithCode(ierror.PreconditionCode).
+				WithMessage("the tag %s with signature cannot be updated", tag.Name)
+		}
+		// the tag already exists under the repository and is attached to the artifact, return directly
+		if tag.ArtifactID == artifactID {
+			return nil
+		}
```

Comment:
Put immutable and signature checking after this, if the tag already exists and is already attached to the same artifact, we shouldn't return error